extension/httpfs/httpfs_curl_client.cpp

```diff
@@ -0,0 +1,486 @@
+			curl_easy_setopt(*curl, CURLOPT_POSTFIELDSIZE, info.buffer_in_len);
+
+			// Apply headers
+			curl_easy_setopt(*curl, CURLOPT_HTTPHEADER, curl_headers ? curl_headers.headers : nullptr);
+
+			res = curl->Execute();
+		}
+
+		curl_easy_getinfo(*curl, CURLINFO_RESPONSE_CODE, &request_info->response_code);
+
+		return TransformResponseCurl(res);
+	}
+
+	unique_ptr<HTTPResponse> Head(HeadRequestInfo &info) override {
+		if (state) {
+			state->head_count++;
+		}
+
+		auto curl_headers = TransformHeadersCurl(info.headers);
+		request_info->url = info.url;
+		// transform parameters
+		if (!info.params.extra_headers.empty()) {
+			auto curl_params = TransformParamsCurl(info.params);
+			request_info->url += "?" + curl_params;
+		}
+
+		CURLcode res;
+		{
+			// Set URL
+			curl_easy_setopt(*curl, CURLOPT_URL, request_info->url.c_str());
+
+			// Perform HEAD request instead of GET
+			curl_easy_setopt(*curl, CURLOPT_NOBODY, 1L);
+			curl_easy_setopt(*curl, CURLOPT_HTTPGET, 0L);
+
+			// Add headers if any
+			curl_easy_setopt(*curl, CURLOPT_HTTPHEADER, curl_headers ? curl_headers.headers : nullptr);
+
+			// Execute HEAD request
+			res = curl->Execute();
+		}
+
+		curl_easy_getinfo(*curl, CURLINFO_RESPONSE_CODE, &request_info->response_code);
+		return TransformResponseCurl(res);
+	}
+
+	unique_ptr<HTTPResponse> Delete(DeleteRequestInfo &info) override {
+		if (state) {
+			state->delete_count++;
+		}
+
+		auto curl_headers = TransformHeadersCurl(info.headers);
+		// transform parameters
+		request_info->url = info.url;
+		if (!info.params.extra_headers.empty()) {
+			auto curl_params = TransformParamsCurl(info.params);
+			request_info->url += "?" + curl_params;
+		}
+
+		CURLcode res;
+		{
+			// Set URL
+			curl_easy_setopt(*curl, CURLOPT_URL, request_info->url.c_str());
+
+			// Set DELETE request method
+			curl_easy_setopt(*curl, CURLOPT_CUSTOMREQUEST, "DELETE");
+
+			// Follow redirects
+			curl_easy_setopt(*curl, CURLOPT_FOLLOWLOCATION, 1L);
+
+			// Add headers if any
+			curl_easy_setopt(*curl, CURLOPT_HTTPHEADER, curl_headers ? curl_headers.headers : nullptr);
+
+			// Execute DELETE request
+			res = curl->Execute();
+		}
+
+		// Get HTTP response status code
+		curl_easy_getinfo(*curl, CURLINFO_RESPONSE_CODE, &request_info->response_code);
+		return TransformResponseCurl( res);
+	}
+
+	unique_ptr<HTTPResponse> Post(PostRequestInfo &info) override {
+		if (state) {
+			state->post_count++;
+			state->total_bytes_sent += info.buffer_in_len;
+		}
+
+		auto curl_headers = TransformHeadersCurl(info.headers);
+		const string content_type = "Content-Type: application/octet-stream";
+		curl_headers.Add(content_type.c_str());
+		// transform parameters
+		request_info->url = info.url;
+		if (!info.params.extra_headers.empty()) {
+			auto curl_params = TransformParamsCurl(info.params);
+			request_info->url += "?" + curl_params;
+		}
+
+		CURLcode res;
+		{
+			curl_easy_setopt(*curl, CURLOPT_URL, request_info->url.c_str());
```

Comment:
curious should we sanity check the return value for option setting?
http://curl.se/libcurl/c/curl_easy_setopt.html